core-active-state.html

```diff
@@ -0,0 +1,175 @@
+<script>
+  Polymer({
+    is: 'core-active-state',
+
+    properties: {
+
+      /**
+       * If true, the element is currently active either because the
+       * user is touching it, or the button is a toggle
+       * and is currently in the active state.
+       *
+       * @attribute active
+       * @type boolean
+       * @default false
+       */
+      active: {
+        type: Boolean,
+        notify: true,
+        readOnly: true,
+        observer: 'activeChanged',
+        value: false
+      },
+
+      /**
+       * If true, the user is currently holding down the button.
+       *
+       * @attribute pressed
+       * @type boolean
+       * @default false
+       */
+      pressed: {
+        type: Boolean,
+        value: function() {
+          this._setPressed(this.pressed || false);
+        },
+        readOnly: true,
+        observer: 'pressedChanged'
+      },
+
+      /**
+       * If true, the button toggles the active state with each tap.
+       * Otherwise, the button becomes active when the user is holding
+       * it down.
+       *
+       * @attribute toggle
+       * @type boolean
+       * @default false
+       */
+      toggle: {
+        type: Boolean,
+        value: false
+      },
+
+      /**
+       * If true, the user cannot interact with this element.
+       *
+       * @attribute disabled
+       * @type boolean
+       * @default false
+       */
+      disabled: {
+        type: Boolean,
+        observer: 'disabledChanged',
+        value: false
+      }
+    },
+
+    get target() {
+      return this.host || this.parentNode;
+    },
+
+    configure: function() {
+      return {
+        boundDownHandler: this.onTargetDown.bind(this),
+        boundUpHandler: this.onTargetUp.bind(this)
+      };
+    },
+
+    ready: function() {
+      this.afterReady = true;
+      // TODO(cdata): We this as a workaround for polymer/polymer#1285
+      if (this.isAttached) {
+        this.attached();
+      }
+    },
+
+    attached: function() {
+      if (!this.afterReady) {
+        return;
+      }
+
+      this.target.addEventListener('mousedown', this.boundDownHandler);
+      this.target.addEventListener('mouseup', this.boundUpHandler);
+      this.target.addEventListener('keydown', this.boundDownHandler);
+      this.target.addEventListener('keyup', this.boundUpHandler);
+    },
+
+    detached: function() {
+      this.target.removeEventListener('mousedown', this.boundDownHandler);
+      this.target.removeEventListener('mouseup', this.boundUpHandler);
+      this.target.removeEventListener('keydown', this.boundDownHandler);
+      this.target.removeEventListener('keyup', this.boundUpHandler);
+    },
+
+    onTargetDown: function(event) {
+      if (event.type === 'keydown' &&
+          event.keyCode !== 32 &&
+          event.keyCode !== 13) {
+        return;
+      }
+
+      if (event.keyCode === 13) {
+        window.setTimeout(function() {
```

Comment:
Ah! Ok, so you're running into the joy that are tasks and microtasks:

(roughly) each entry into the JS VM occurs as a task (a "tick" of the run loop). Each task also maintains a queue of _microtasks_, which are callbacks that will be executed at the end of the current task, but before the task returns control back to the event loop.

`process.nextTick` (node), mutation observers, (native) promises, object.observe callbacks queue microtasks.

`setTimeout`, `setImmediate`, `setInterval`, `requestAnimationFrame`, etc queue up tasks.

---

You also need to be careful about `setTimeout`, due to its clamping behavior (which is also browser-dependent). In Chrome, for example, the first few calls to `setTimeout(..., 0)` per stack occur as immediate tasks, but subsequent calls are clamped to a 4ms delay.

You should be ok here, as this is being triggered as part of an event handler (e.g. a new stack).

---

However, if you want the click event to fire after `mouseup`/`keyup`, you should be performing this logic in `onTargetUp`, not `onTargetDown`!

The only thing you can guarantee here is that `click` will be called after the down events